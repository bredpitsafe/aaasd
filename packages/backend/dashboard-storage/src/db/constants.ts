export enum ETable {
    Dashboards = 'dashboards',
    DashboardLog = 'dashboard_log',
    Permissions = 'permissions',
    PermissionLog = 'permission_log',
    Scopes = 'scopes',
    ScopesHistory = 'scopes_history',
    Drafts = 'drafts',
}
