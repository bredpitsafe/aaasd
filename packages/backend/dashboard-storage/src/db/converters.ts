import { camelCase, mapKeys, mapValues, snakeCase } from 'lodash-es';

import type { TDbDashboard, TGrpcDashboard } from '../def/dashboard.ts';
import { toGrcpKind } from './enum-convertors/kind.convertor.ts';
import { toGrcpStatus } from './enum-convertors/status.convertor.ts';

/* eslint-disable */
export function fromDB<T extends Record<string, any>>(obj: T): T {
    const mappedValues = mapValues(obj, (v) => v) as T;
    return mapKeys(mappedValues, (_, k) => camelCase(k)) as T;
}

export function toDB<T extends Record<string, any>>(obj: T): Record<string, any> {
    return mapKeys(obj, (_, k) => snakeCase(k));
}

export const fromDbDashboard = (dbDashboard: TDbDashboard): TGrpcDashboard => ({
    ...dbDashboard,
    kind: toGrcpKind(dbDashboard.kind),
    status: toGrcpStatus(dbDashboard.status),
});
