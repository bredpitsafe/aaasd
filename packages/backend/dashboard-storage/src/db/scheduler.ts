import { generateTraceId } from '@common/utils';
import Bottleneck from 'bottleneck';

import { EActorName } from '../def/actor.ts';
import { config } from '../utils/config.ts';
import { defaultLogger } from '../utils/logger.ts';

export const queryLimiter = new Bottleneck({
    maxConcurrent: config.postgres.maxConcurrentQueries,
    highWater: config.postgres.maxQueueSize,
    strategy: Bottleneck.strategy.OVERFLOW,
});

queryLimiter.on('failed', function (error, jobInfo) {
    // This will be called every time a job fails.
    defaultLogger.error({
        actor: EActorName.Root,
        message: 'limiter job failed',
        traceId: generateTraceId(),
        error,
        jobInfo,
    });
});
