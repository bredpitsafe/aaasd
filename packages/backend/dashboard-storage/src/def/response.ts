/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @deprecated
 */
export type DashboardItem = DashboardsListItem & {
  draftDigest: string;
};


// TODO: remove this file after enums transfer into more appropriate place and tests fix
/**
 * @deprecated
 */
export interface Response {
  error?: ErrorResponse;
  payload?:
    | UnsubscribeResponse
    | PingResponse
    | HeartbeatResponse
    | AuthenticateResponse
    | LogoutResponse
    | SubscribeToDashboardResponse
    | SubscribeToDashboardsListResponse
    | FetchDashboardConfigResponse
    | FetchDashboardDraftResponse
    | CreateDashboardResponse
    | ImportDashboardResponse
    | UpdateDashboardResponse
    | UpdateDashboardDraftResponse
    | ResetDashboardDraftResponse
    | UpdateDashboardPermissionsResponse
    | UpdateDashboardShareSettingsResponse
    | DeleteDashboardResponse
    | SubscribeToDashboardPermissionsResponse
    | FetchDashboardIdByLegacyIdResponse
    | RenameDashboardResponse
    | UpdateDashboardScopeBindingResponse;
  timestamp: string;
  traceId: string;
  correlationId: number;
  state: ResponseState;
}
/**
 * @deprecated
 */
export interface ErrorResponse {
  kind: string;
  description: string;
  args?: {
    [k: string]: unknown;
  };
}
/**
 * @deprecated
 */
export interface UnsubscribeResponse {
  type: "Unsubscribed";
}
/**
 * @deprecated
 */
export interface PingResponse {
  type: "Pong";
}
/**
 * @deprecated
 */
export interface HeartbeatResponse {
  type: "Heartbeat";
}
/**
 * @deprecated
 */
export interface AuthenticateResponse {
  type: "Authenticated";
  username: string;
}
/**
 * @deprecated
 */
export interface LogoutResponse {
  type: "LoggedOut";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardResponse {
  type: "Dashboard";
  dashboard: DashboardItem;
}
/**
 * @deprecated
 */
export interface DashboardsListItem {
  id: string;
  name: string;
  kind: Kind;
  status: Status;
  digest?: string;
    // This is update time actually
  insertionTime: string;
  hasDraft: boolean;
  sharePermission: SharePermission;
  permission: Permission;
  owners: User[];
  permissionsCount?: number;
  scopes: Scope[];
}
/**
 * @deprecated
 */
export interface User {
  user: string;
}
/**
 * @deprecated
 */
export interface Scope {
  [k: string]: string | number | boolean | null;
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardsListResponse {
  type: "DashboardsList";
  list: DashboardsListItem[];
}
/**
 * @deprecated
 */
export interface FetchDashboardConfigResponse {
  type: "DashboardConfig";
  config: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface FetchDashboardDraftResponse {
  type: "DashboardDraft";
  draft: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface CreateDashboardResponse {
  type: "DashboardCreated";
  id: string;
}
/**
 * @deprecated
 */
export interface ImportDashboardResponse {
  type: "DashboardImported";
  id: string;
}
/**
 * @deprecated
 */
export interface UpdateDashboardResponse {
  type: "DashboardUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardDraftResponse {
  type: "DashboardDraftUpdated";
}
/**
 * @deprecated
 */
export interface ResetDashboardDraftResponse {
  type: "DashboardDraftReset";
}
/**
 * @deprecated
 */
export interface UpdateDashboardPermissionsResponse {
  type: "PermissionsUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardShareSettingsResponse {
  type: "DashboardShareSettingsUpdated";
}
/**
 * @deprecated
 */
export interface DeleteDashboardResponse {
  type: "DashboardDeleted";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardPermissionsResponse {
  type: "DashboardPermissionsList";
  list: {
    user: string;
    permission: string;
  }[];
}
/**
 * @deprecated
 */
export interface FetchDashboardIdByLegacyIdResponse {
  type: "DashboardId";
  id: string;
}
/**
 * @deprecated
 */
export interface RenameDashboardResponse {
  type: "DashboardRenamed";
}
/**
 * @deprecated
 */
export interface UpdateDashboardScopeBindingResponse {
  type: "DashboardScopeBindingUpdated";
  platformTime: string;
}

// TODO: move this enums to the approriate place
export const enum Kind {
  User = "User",
  Robot = "Robot"
}
export const enum Status {
  Active = "Active",
  Suspended = "Suspended",
  Archived = "Archived"
}
export const enum SharePermission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor"
}
export const enum Permission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor",
  Owner = "Owner"
}
/**
 * @deprecated
 */
export const enum ResponseState {
  Done = "Done",
  InProgress = "InProgress"
}
