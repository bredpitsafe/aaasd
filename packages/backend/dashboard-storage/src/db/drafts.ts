import type { TUserName } from '@common/types/src/primitives/index.ts';
import { isEmpty } from 'lodash-es';
import type { Observable } from 'rxjs';
import { switchMap } from 'rxjs';

import type { TDbDashboard } from '../def/dashboard.ts';
import type {
    TDashboardDraftLookup,
    TDbDashboardDraft,
    TUpdatableDbDashboardDraft,
} from '../def/draft.ts';
import { ETable } from './constants.ts';
import { client } from './postgres/index.ts';
import { subscribeToTableUpdates } from './subscription.ts';

type TGetDashboardDraftsLookupQuery = {
    user: TUserName;
    ids: TDbDashboard['id'][];
};

export async function upsertDraft(draft: TUpdatableDbDashboardDraft): Promise<void> {
    await client.insert<TUpdatableDbDashboardDraft>({
        table: ETable.Drafts,
        values: [draft],
        upsert: '"user", dashboard_id',
    });
}

export async function deleteDraft(
    draft: Pick<TUpdatableDbDashboardDraft, 'user' | 'dashboardId'>,
): Promise<void> {
    await client.insert<TUpdatableDbDashboardDraft>({
        table: ETable.Drafts,
        values: [{ ...draft, config: '', digest: '' }],
        upsert: '"user", dashboard_id',
    });
}

async function getCommonDrafts<T extends TDbDashboardDraft | TDashboardDraftLookup>(
    fields: string,
    user: TUserName,
    ids: TDbDashboard['id'][],
): Promise<T[]> {
    if (isEmpty(ids)) {
        return [];
    }

    return client.query<T>({
        query: `SELECT ${fields} from ${ETable.Drafts} WHERE "user" = {user: String} AND dashboard_id IN ({ids: Array(String)}) AND config != ''`,
        query_params: {
            user,
            ids,
        },
    });
}

function getPartialDrafts(
    user: TUserName,
    dashboardIds: TDbDashboard['id'][],
): Promise<TDashboardDraftLookup[]> {
    return getCommonDrafts('dashboard_id, digest', user, dashboardIds);
}

export async function getDraftById(
    user: TUserName,
    id: TDbDashboard['id'],
): Promise<TDbDashboardDraft | undefined> {
    const [draft] = await getCommonDrafts<TDbDashboardDraft>('*', user, [id]);
    return draft;
}

const trigger$ = subscribeToTableUpdates('drafts_update');

export function subscribeToDashboardDraftLookupUpdates(
    query: TGetDashboardDraftsLookupQuery,
): Observable<TDashboardDraftLookup[]> {
    return trigger$.pipe(switchMap(() => getPartialDrafts(query.user, query.ids)));
}
