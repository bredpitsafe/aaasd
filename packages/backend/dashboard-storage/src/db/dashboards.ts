import { randomUUID } from 'crypto';
import { first, isEmpty, isEqual } from 'lodash-es';
import type { Observable } from 'rxjs';
import { distinctUntilChanged, switchMap } from 'rxjs';

import type { TDbDashboard, TGrpcDashboard } from '../def/dashboard.ts';
import { ETable } from './constants.ts';
import { fromDbDashboard } from './converters.ts';
import { fromGrpcKind } from './enum-convertors/kind.convertor.ts';
import { fromGrpcStatus } from './enum-convertors/status.convertor.ts';
import { client } from './postgres/index.ts';
import { subscribeToTableUpdates } from './subscription.ts';
import { getDigest } from './utils.ts';

type TUpdatableDbDashboard = Pick<TDbDashboard, 'name' | 'config'> &
    Partial<Omit<TDbDashboard, 'name' | 'config'>>;

export async function upsertDashboard(
    dashboard: Pick<TGrpcDashboard, 'name' | 'config'> &
        Partial<Omit<TGrpcDashboard, 'name' | 'config'>>,
): Promise<TDbDashboard['id']> {
    const id = dashboard.id ?? randomUUID();

    const newDashboard: TUpdatableDbDashboard = {
        id,
        config: dashboard.config,
        name: dashboard.name,
        kind: fromGrpcKind(dashboard.kind ?? 'KIND_USER'),
        status: fromGrpcStatus(dashboard.status ?? 'STATUS_ACTIVE'),
        digest: getDigest(dashboard.config),
    };

    await client.insert<TUpdatableDbDashboard>({
        table: ETable.Dashboards,
        values: [newDashboard],
        upsert: 'id',
    });

    return id;
}

export async function getDashboardById(
    id: TGrpcDashboard['id'],
): Promise<TGrpcDashboard | undefined> {
    const result = await client.query<TDbDashboard>({
        query: `SELECT * from ${ETable.Dashboards} WHERE id = {id: String}`,
        query_params: {
            id,
        },
    });

    const dbDashboard = first(result);

    return dbDashboard && fromDbDashboard(dbDashboard);
}

export async function getDashboardsByDigest(
    digest: TGrpcDashboard['digest'],
): Promise<Pick<TGrpcDashboard, 'id'>[]> {
    return client.query<Pick<TDbDashboard, 'id'>>({
        query: `SELECT id from ${ETable.Dashboards} WHERE digest = {digest: String}`,
        query_params: {
            digest,
        },
    });
}

export async function getDashboardAtRevision(
    id: TGrpcDashboard['id'],
    digest: TGrpcDashboard['digest'],
): Promise<TGrpcDashboard | undefined> {
    const result = await client.query<TDbDashboard>({
        query: `SELECT * from ${ETable.DashboardLog} WHERE id = {id: String} AND digest = {digest: String} LIMIT 1`,
        query_params: {
            id,
            digest,
        },
    });

    const dbDashboard = first(result);

    return dbDashboard && fromDbDashboard(dbDashboard);
}

async function getCommonDashboards(
    fields: string,
    ids: TGrpcDashboard['id'][],
): Promise<TGrpcDashboard[]> {
    if (isEmpty(ids)) {
        return [];
    }

    const dbDashboards = await client.query<TDbDashboard>({
        query: `SELECT ${fields} from ${ETable.Dashboards} WHERE id IN ({ids: Array(String)})`,
        query_params: {
            ids,
        },
    });

    return dbDashboards.map((dbDashboard) => fromDbDashboard(dbDashboard));
}

const trigger$ = subscribeToTableUpdates('dashboards_update');

type TGetDashboardsQuery = {
    ids: TGrpcDashboard['id'][];
};

export function subscribeToDashboardUpdates(
    query: TGetDashboardsQuery,
): Observable<TGrpcDashboard[]> {
    return trigger$.pipe(
        switchMap(() =>
            getCommonDashboards('id,name,kind,status,insertion_time,digest', query.ids),
        ),
        distinctUntilChanged<TGrpcDashboard[]>(isEqual),
    );
}
