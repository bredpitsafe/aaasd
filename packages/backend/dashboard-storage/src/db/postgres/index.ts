import type { TraceId } from '@common/utils';
import type { ClientConfig, PoolConfig, QueryResultRow } from 'pg';
import pg from 'pg';
import format from 'pg-format';
import type { Observable } from 'rxjs';
import { catchError, finalize, from, fromEvent, map, of, switchMap, throwError } from 'rxjs';
import { tap } from 'rxjs/operators';

import { EActorName } from '../../def/actor.ts';
import { EMetricsLabel } from '../../def/metrics.ts';
import { config } from '../../utils/config.ts';
import { ActorError, EActorErrorKind } from '../../utils/errors.ts';
import { defaultLogger } from '../../utils/logger.ts';
import { metrics } from '../../utils/metrics.ts';
import { fromDB, toDB } from '../converters.ts';
import { operation } from './utils.ts';

const clientConfig: ClientConfig = {
    host: config.postgres.host,
    port: config.postgres.port,
    database: config.postgres.databaseName,
    user: config.postgres.username,
    password: config.postgres.password,
    keepAlive: true,
    application_name: config.service.name,
};

const poolConfig: PoolConfig = {
    ...clientConfig,
    min: config.postgres.minPoolSize,
    max: config.postgres.maxPoolSize,
    allowExitOnIdle: config.postgres.allowExitOnIdle,
    connectionTimeoutMillis: config.postgres.connectionTimeout,
    idleTimeoutMillis: config.postgres.idleTimeout,
};

const pool = new pg.Pool(poolConfig);

const updatePoolMetrics = () => {
    metrics.db.pool.total.set(pool.totalCount);
    metrics.db.pool.idle.set(pool.idleCount);
    metrics.db.pool.waiting.set(pool.waitingCount);
};

const dbLogger = defaultLogger.createChildLogger({ actor: EActorName.Database });

pool.on('error', () => {
    // Do nothing, error will be caught by pool clients.
    // Only catch error here since without catching it `pg` will throw global unhandled exception
});
pool.on('connect', updatePoolMetrics);
pool.on('acquire', updatePoolMetrics);
pool.on('release', updatePoolMetrics);
pool.on('remove', updatePoolMetrics);

type InsertParams<T> = {
    table: string;
    values: T[];
    upsert?: string;
};

const insert = async <T extends Record<string, unknown>>(params: InsertParams<T>) => {
    if (params.values.length === 0) {
        throw new Error('insert values list is empty');
    }

    const values = params.values.map(toDB);
    const [value] = values;
    const keys = Object.keys(value);
    const conflictKeys = params.upsert ? keys.map((key) => [key, key]).flat() : [];

    const items = values.map((v) => Object.values(v) as unknown[]);
    const sqlTemplate = `INSERT INTO %I(%I) VALUES ${Array(items.length).fill('(%L)').join(', ')}${
        params.upsert
            ? ` ON CONFLICT (%s) DO UPDATE SET ${keys.map(() => '%I = excluded.%I').join(', ')}`
            : ''
    }`;
    const sql = format(sqlTemplate, params.table, keys, ...items, params.upsert, ...conflictKeys);

    return operation('insert', sql, pool, params.table);
};

type TDeleteParams<T extends Record<string, unknown>> = {
    table: string;
    conditions: Partial<T>;
};

const deleteRows = async <T extends Record<string, unknown>>(params: TDeleteParams<T>) => {
    const keys = Object.keys(toDB(params.conditions));
    const values = Object.values(params.conditions);

    if (keys.length === 0) {
        throw new Error('Table name and conditions are required for deletion.');
    }

    const whereClause = keys
        .map((key, index) => format('%I = %L', key, values[index]))
        .join(' AND ');

    const sqlTemplate = `DELETE FROM %I WHERE ${whereClause}`;
    const sql = format(sqlTemplate, params.table);

    return operation('delete', sql, pool, params.table);
};

type TQueryParams = {
    query: string;
    query_params?: Record<string, unknown>;
};

const query = async <T extends QueryResultRow>(params: TQueryParams): Promise<T[]> => {
    // Find replacement matches in the query (format is similar to Clickhouse client)
    let sql = params.query;
    const re = /\{(\w+):.+?\}/g;
    let match: RegExpExecArray | null;
    const matches: { key: string; originalStr: string }[] = [];
    while ((match = re.exec(params.query)) != null) {
        matches.push({ key: match[1], originalStr: match[0] });
    }

    // Replace matches in query
    const values = matches.reduce((acc, match) => {
        sql = sql.replaceAll(match.originalStr, `%L`);
        const value = params.query_params?.[match.key];
        acc.push(value);
        return acc;
    }, [] as unknown[]);

    // Format query with replacement values
    sql = format(sql, ...values);

    const { rows } = await operation<T>('query', sql, pool, params.query);

    return rows.map(fromDB);
};

type TListenParams = {
    eventName: string;
    traceId: TraceId;
};

const listen = <T>(params: TListenParams): Observable<T> => {
    const { eventName, traceId } = params;
    return of(undefined).pipe(
        switchMap(() => pool.connect()),
        tap(() => {
            metrics.db.watchQueriesActive.inc({ [EMetricsLabel.Type]: eventName });
            dbLogger.debug({
                message: 'subscribed to notification events',
                eventName,
                traceId,
            });
        }),
        switchMap((client) => {
            const query = `LISTEN "${eventName}"`;
            return from(client.query(query)).pipe(
                switchMap(() => {
                    dbLogger.debug({
                        message: 'LISTEN query finished, notifications are active',
                        eventName,
                        traceId,
                    });

                    return fromEvent<{ payload: string }>(client, 'notification');
                }),
                tap((event) =>
                    dbLogger.debug({
                        message: 'received notification event',
                        eventName,
                        event,
                        traceId,
                    }),
                ),
                map((event) => JSON.parse(event.payload) as T),

                finalize(() => {
                    metrics.db.watchQueriesActive.dec({ [EMetricsLabel.Type]: eventName });
                    void client.release(true);
                    dbLogger.debug({
                        message:
                            'unsubscribed from notification events & released notifications client back to the pool',
                        eventName,
                        traceId,
                    });
                }),
            );
        }),
        catchError((err) => {
            return throwError(
                () =>
                    new ActorError({
                        kind: EActorErrorKind.Database,
                        title: 'Database notification subscription has failed',
                        description: err.message,
                    }),
            );
        }),
    );
};

export const client = { insert, delete: deleteRows, query, listen };
