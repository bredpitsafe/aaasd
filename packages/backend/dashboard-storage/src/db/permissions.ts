import type { TUserName } from '@common/types/src/primitives/index.ts';
import type { TPermission } from '@grpc-schemas/dashboard_storage-api-sdk/services/dashboard_storage/v1/common';
import { isEmpty, isEqual, isNil } from 'lodash-es';
import { distinctUntilChanged, map, switchMap } from 'rxjs';

import type { TDbDashboard, TGrpcDashboard } from '../def/dashboard.ts';
import type { TDbPermissionRow, TGrpcPermissionRow } from '../def/permissions.ts';
import { EGroups } from '../def/permissions.ts';
import { Permission } from '../def/response.ts';
import { ETable } from './constants.ts';
import { fromGrpcPermission, toGrcpPermission } from './enum-convertors/permission.convertor.ts';
import { client } from './postgres/index.ts';
import { subscribeToTableUpdates } from './subscription.ts';

type TGetPermissionsQuery = {
    user: TUserName;
    id?: TDbDashboard['id'];
    allowedLevels?: TPermission[];
};

type TResetPermissionsQuery = {
    user?: TUserName;
    id: TDbDashboard['id'];
};

type TGetDashboardPermissionsQuery = {
    user?: TUserName;
    id: TDbDashboard['id'];
};

type TGetDashboardsOwnersQuery = {
    ids: TDbDashboard['id'][];
};

type TGetPermissionsCountQuery = {
    ids: TDbDashboard['id'][];
    user: TUserName;
};

type TUpdatableDbPermissionRow = Pick<TDbPermissionRow, 'user' | 'permission'>;
export type TUpdatableGrpcPermissionRow = Pick<TGrpcPermissionRow, 'user' | 'permission'>;

export async function getPermissions(query: TGetPermissionsQuery): Promise<TGrpcPermissionRow[]> {
    const chQuery = `SELECT dashboard_id, "user", permission, insertion_time from ${
        ETable.Permissions
    } WHERE "user" in ({users: Array(String)})${
        query.id ? ` AND dashboard_id = {id: String}` : ''
    }${
        !(isNil(query.allowedLevels) || isEmpty(query.allowedLevels))
            ? ` AND permission in ({allowedLevels: Array(String)})`
            : ''
    } ORDER BY dashboard_id, "user" ASC;`;

    const dbPermissionRows = await client.query<TDbPermissionRow>({
        query: chQuery,
        query_params: {
            id: query.id ?? '',
            users: [query.user, EGroups.All],
            allowedLevels: query.allowedLevels?.map((level) => fromGrpcPermission(level)) ?? [],
        },
    });

    return dbPermissionRows.map((dbPermissionRow) => ({
        ...dbPermissionRow,
        permission: toGrcpPermission(dbPermissionRow.permission),
    }));
}

export async function getDashboardPermissions(
    query: TGetDashboardPermissionsQuery,
): Promise<TUpdatableGrpcPermissionRow[]> {
    const chQuery = `SELECT "user", permission from ${ETable.Permissions} WHERE permission != '${
        Permission.None
    }' AND dashboard_id = {id: String}${
        !isNil(query.user) ? ' AND "user" != {user: String}' : ''
    } AND "user" != '${EGroups.All}' ORDER BY "user" ASC`;

    const dbPermissionRows = await client.query<TDbPermissionRow>({
        query: chQuery,
        query_params: {
            id: query.id,
            user: query.user ?? '',
        },
    });

    return dbPermissionRows.map((dbPermissionRow) => ({
        ...dbPermissionRow,
        permission: toGrcpPermission(dbPermissionRow.permission),
    }));
}

async function getDashboardsOwners(
    query: TGetDashboardsOwnersQuery,
): Promise<Pick<TDbPermissionRow, 'user' | 'dashboardId'>[]> {
    const chQuery = `SELECT "user", dashboard_id from ${ETable.Permissions} FINAL WHERE permission = {permission: String} AND dashboard_id IN ({ids: Array(String)}) ORDER BY "user" ASC`;

    return client.query<TDbPermissionRow>({
        query: chQuery,
        query_params: {
            ids: query.ids,
            permission: Permission.Owner,
        },
    });
}

export async function resetDashboardPermissions(query: TResetPermissionsQuery): Promise<void> {
    const permissionRows = await getDashboardPermissions(query);
    const resetPermissionRows: TUpdatableGrpcPermissionRow[] = permissionRows.map(
        (permissionRow) => ({
            user: permissionRow.user,
            permission: 'PERMISSION_NONE',
        }),
    );
    await upsertPermissions(query.id, resetPermissionRows);
}

export async function upsertPermissions(
    dashboardId: TGrpcDashboard['id'],
    grpcPermissionRows: TUpdatableGrpcPermissionRow[],
) {
    const newDbPermissionRows = grpcPermissionRows.map((grpcPermissionRow) => ({
        ...grpcPermissionRow,
        permission: fromGrpcPermission(grpcPermissionRow.permission),
        dashboardId,
    }));

    await client.insert<TUpdatableDbPermissionRow>({
        table: ETable.Permissions,
        values: newDbPermissionRows,
        upsert: '"user", dashboard_id',
    });

    return dashboardId;
}

function getPermissionsCount(
    query: TGetPermissionsCountQuery,
): Promise<{ dashboardId: TDbDashboard['id']; count: string }[]> {
    const chQuery = `SELECT count(DISTINCT "user") as count, dashboard_id from ${ETable.Permissions} FINAL WHERE dashboard_id in ({ids: Array(String)}) AND permission != '${Permission.None}' AND "user" != {user: String} AND "user" != '${EGroups.All}' GROUP BY dashboard_id`;

    return client.query({
        query: chQuery,
        query_params: {
            ids: query.ids,
            user: query.user,
        },
    });
}

const trigger$ = subscribeToTableUpdates('permissions_update');

export function subscribeToPermissionsUpdates(query: TGetPermissionsQuery) {
    return trigger$.pipe(
        switchMap(() => getPermissions(query)),
        distinctUntilChanged<TGrpcPermissionRow[]>(isEqual),
    );
}

export function subscribeToDashboardPermissionsUpdates(query: TGetDashboardPermissionsQuery) {
    return trigger$.pipe(
        switchMap(() => getDashboardPermissions(query)),
        distinctUntilChanged<TUpdatableGrpcPermissionRow[]>(isEqual),
    );
}

export function subscribeToPermissionsCountUpdates(query: TGetPermissionsCountQuery) {
    return trigger$.pipe(
        switchMap(() => getPermissionsCount(query)),
        map(
            (items) =>
                items.reduce((acc: Record<TDbDashboard['id'], { count: number }>, item) => {
                    acc[item.dashboardId] = { count: Number(item.count) };
                    return acc;
                }, {}),
            distinctUntilChanged<Record<TDbDashboard['id'], { count: number }>>(isEqual),
        ),
    );
}

export function subscribeToDashboardsOwnersUpdates(query: TGetDashboardsOwnersQuery) {
    return trigger$.pipe(
        switchMap(() => getDashboardsOwners(query)),
        map(
            (items) =>
                items.reduce((acc: Record<TDbPermissionRow['dashboardId'], TUserName[]>, item) => {
                    const { user, dashboardId: id } = item;
                    if (isNil(acc[id])) {
                        acc[id] = [];
                    }
                    acc[id].push(user);
                    return acc;
                }, {}),
            distinctUntilChanged<Record<TDbPermissionRow['dashboardId'], TUserName[]>>(isEqual),
        ),
    );
}
