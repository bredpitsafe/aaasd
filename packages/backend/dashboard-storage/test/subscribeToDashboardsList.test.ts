import { wait } from '@common/utils/src/wait.js';
import { firstValueFrom, skip } from 'rxjs';
import { afterAll, beforeEach, describe, expect, it, vi } from 'vitest';

import type {
    CreateDashboardResponse,
    SubscribeToDashboardsListResponse,
} from '../src/def/response.ts';
import { Permission } from '../src/def/response.ts';
import { Client, users } from './client/index.js';
import initialDashboard from './fixtures/dashboard.json';
import { clearDashboards } from './utils/db.js';

const LIST_SUBSCRIPTION_QUERY = { type: 'SubscribeToDashboardsList' };

describe('SubscribeToDashboardsList', () => {
    const client = new Client({ authenticate: true });
    const otherClient = new Client({ authenticate: true, as: users.user2.username });

    beforeEach(async () => {
        await clearDashboards();
    });

    afterAll(() => {
        client.close();
        otherClient.close();
    });

    it('gets empty list when no dashboards in DB', async () => {
        const response = await client.request(LIST_SUBSCRIPTION_QUERY);
        expect(response.state).toBe('InProgress');
        expect(response.payload).toMatchSnapshot();
    });

    it('gets dashboard in list snapshot after creation', async () => {
        const { id } = (await client.request(initialDashboard)).payload;
        const response = await client.request(LIST_SUBSCRIPTION_QUERY);
        expect(response.state).toBe('InProgress');
        expect(response.payload.list).toHaveLength(1);
        expect(response.payload.list[0]).toMatchObject({
            id,
            name: initialDashboard.name,
            hasDraft: false,
            digest: expect.any(String),
            permission: 'Owner',
        });
    });

    it('updates the list after adding a new dashboard', async () => {
        // Step 1. Create dashboard 1.
        await client.request(initialDashboard);

        // Step 2. Subscribe to list, skip snapshot and wait for the update.
        const responsePromise = firstValueFrom(client.send(LIST_SUBSCRIPTION_QUERY).pipe(skip(1)));

        // Step 3. Create dashboard 2.
        const { id } = (await client.request(initialDashboard)).payload as CreateDashboardResponse;

        const response = await responsePromise;
        const payload = response.payload as SubscribeToDashboardsListResponse;
        expect(response.state).toBe('InProgress');
        expect(payload.list).toHaveLength(2);
        expect(payload.list[1]).toMatchObject({
            id,
            name: initialDashboard.name,
            hasDraft: false,
            digest: expect.any(String),
            permission: 'Owner',
        });
    });

    it('updates the list after adding a dashboard through permissions', async () => {
        // Step 1. Subscribe to list, skip snapshot and wait for the update.
        const responsePromise = firstValueFrom(client.send(LIST_SUBSCRIPTION_QUERY).pipe(skip(1)));

        // Step 2. Create dashboard as another user.
        const { id } = (await otherClient.request(initialDashboard)).payload;

        // Step 3. Grant `Editor` permissions to the new dashboard to user1.
        const permission = Permission.Editor;
        await otherClient.request({
            type: 'UpdateDashboardPermissions',
            id,
            permissions: [{ user: client.username, permission }],
        });

        // Step 4. Wait for an update to come when permission has been granted
        const response = await responsePromise;
        const payload = response.payload as SubscribeToDashboardsListResponse;
        expect(response.state).toBe('InProgress');
        expect(payload.list).toHaveLength(1);
        expect(payload.list[0]).toMatchObject({
            id,
            name: initialDashboard.name,
            hasDraft: false,
            digest: expect.any(String),
            permission,
        });
    });

    it('updates the list after removing dashboard permissions', async () => {
        // Step 1. Create dashboard as another user and grant `Owner` permissions to user1.
        const { id } = (await otherClient.request(initialDashboard)).payload;

        await otherClient.request({
            type: 'UpdateDashboardPermissions',
            id,
            permissions: [{ user: client.username, permission: Permission.Owner }],
        });

        // Step 3. Subscribe to the list and wait for an update.
        const responsePromise = firstValueFrom(client.send(LIST_SUBSCRIPTION_QUERY).pipe(skip(1)));

        // Step 4. Revoke previously assigned permissions.
        await otherClient.request({
            type: 'UpdateDashboardPermissions',
            id,
            permissions: [{ user: client.username, permission: Permission.None }],
        });

        // Step 4. Wait for an update to come when permission has been granted
        const response = await responsePromise;
        const payload = response.payload as SubscribeToDashboardsListResponse;
        expect(response.state).toBe('InProgress');
        expect(payload.list).toHaveLength(0);
    });

    it('updates the list when owners change', async () => {
        // Step 1. Create dashboard.
        const { id } = (await client.request(initialDashboard)).payload;

        // Step 2. Subscribe to dashboards list
        const subscriber = vi.fn();
        const subscription$ = client.send(LIST_SUBSCRIPTION_QUERY);
        const subscription = subscription$.subscribe(subscriber);

        await wait();

        // Step 3. Add another owner
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Owner' }],
        };
        await client.request(permission);
        await wait();

        // Step 4. Check that update has been received
        const { length } = subscriber.mock.calls;
        expect(length).toBeGreaterThanOrEqual(2);
        expect(subscriber.mock.calls[0][0].payload.list[0].owners).toEqual([
            { user: client.username },
        ]);
        expect(subscriber.mock.calls[length - 1][0].payload.list[0].owners).toEqual([
            { user: client.username },
            { user: otherClient.username },
        ]);

        subscription.unsubscribe();
    });

    it('updates permissionsCount when dashboard is shared among other users', async () => {
        // Step 1. Create dashboard.
        const { id } = (await client.request(initialDashboard)).payload;

        // Step 2. Subscribe to dashboards list
        const subscriber = vi.fn();
        const subscription$ = client.send(LIST_SUBSCRIPTION_QUERY);
        const subscription = subscription$.subscribe(subscriber);

        await wait();

        // Step 3. Share dashboard with another user
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Viewer' }],
        };
        await client.request(permission);
        await wait(1000);

        // Step 4. Check that permissionsCount has changed
        expect(subscriber.mock.calls).toHaveLength(2);
        expect(subscriber.mock.calls[0][0].payload.list[0].permissionsCount).toBe(0);
        expect(subscriber.mock.calls[1][0].payload.list[0].permissionsCount).toBe(1);

        subscription.unsubscribe();
    });
});
