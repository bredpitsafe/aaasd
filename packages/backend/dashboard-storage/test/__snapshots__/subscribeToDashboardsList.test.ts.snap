// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubscribeToDashboardsList > gets empty list when no dashboards in DB 1`] = `
{
  "list": [],
  "type": "DashboardsList",
}
`;
